### ISSUE TITLE
Wheels for Linux, Mac

### ISSUE THREAD

**Issue by matthew-brett** [Roche/pyreadstat#10]

**Comment by matthew-brett** [comment]
I made a repository to build wheels for Manylinux and Mac, then upload them to this Rackspace container:

https://3f23b170c54c2533c070-1c8a9b3114517dc5fe17b7c3f8c63a43.ssl.cf2.rackcdn.com/

The repository is here, with instructions in the README.

I've invited @ofajardo as a collaborator, happy to add others (@ofajardo is an admin and can do that too).

Would you consider uploading the built wheels to PyPI?  I"m happy to do that too, if it's easier.   My PyPI username is `matthew.brett` . Here's [my profile on PyPI](https://pypi.org/user/matthew.brett).

I was less successful on Windows, I think the `readstat` code doesn't compile with MSVC:

https://ci.appveyor.com/project/matthew-brett/pyreadstat-wheels

**Comment by matthew-brett** [comment]
Sorry - repository is here:

https://github.com/MacPython/pyreadstat-wheels

**Comment by ofajardo** [comment]
Hi!

Thanks a lot for doing this!

So, just one question regarding building on windows. Indeed Readstat currently does not compile with MSVC, so I am compiling and building the builds with MINGW (the process is a bit hacky and it is described [here](https://github.com/Roche/pyreadstat/blob/master/windows_compilation.md)).

Do you think is there any chance to get that or some variation of that to work on appveyor?  If not, I definitely want to produce windows wheels, therefore what could be the best solution? I guess I produce them as usual and then manually copy and paste in the same folder with your wheels before pushing to PyPi?

Another minor question: what is the issue with python 3.4 and pandas? I think it's working in my hands (test pass), but probably I am not using the newest version of pandas.

thanks again!

**Comment by matthew-brett** [comment]
Yes, I'm sure it's possible to get something working on Appveyor.

Does conda ship a library for `readstat`?  In that case we might be able to install and statically link to that.

I think it's just the current version of Pandas that doesn't have a 3.4 wheel.    If you would like to compile 3.4 wheels for pyreadstat, then we could install an earlier version of Pandas.   It's true that many projects are dropping 3.4, and even 3.5, these days.

**Comment by ofajardo** [comment]
OK, let's forget about 3.4. What about 2.7, do you think it is still good to support it?

Unfortunately I don't think conda ships readstat. The only project using readstat is R's Haven, but they do what I do, meaning copying manually the source files into their project (well, I mirrored that strategy from them).

I am not sure how to proceed. You have more experience on this deployment topic, so what do you suggest?

**Comment by matthew-brett** [comment]
Ouch - I've read the doc now, about Windows builds, it sounds rough.   I'll ask around to see if anyone has a good setup for this.

In the meantime, OK to upload the Mac / Linux wheels?  I think these are maintainable.

Is there a problem with 2.7?   I guess it will be around for a while, so good to support it if possible.

**Comment by ofajardo** [comment]
Yes, let's put your mac and linux wheels on pypi, thats already very good! 

I have introduced some changes today and bumped the version. On Monday I'm going to build the windows wheels (2.7, 3.5, 3.6 ,3.7) and put them in the repo. After that or in parallel you can do your builds. We join them and they go to pypi, for win mac and linux. Sounds good? I'll make you collaborator on Pypi so that you can push.

**Comment by matthew-brett** [comment]
Hi - thanks - yes, sounds good.

The Rackspace container is just a staging area for uploads to PyPI (usually, at least) - so no need for the Windows wheels there, if they are on PyPI.

Yes, happy to build the wheels for the next release.   It would be good to have a tag for the release, when ready, to make it easier to see which commit is the release.  I don't think there is one for 0.2.0.

Yes, I will keep an eye out for Windows improvements.  I've emailed the author of the previous Python / mingw toolchain, to ask him whether it is practical yet, to build the Python 3 wheels using his modifications.

**Comment by ofajardo** [comment]
Thanks for looking into the windows issue.

Another thing I realized is that Readstat is built on appyevor using mingw

https://github.com/WizardMac/ReadStat/blob/master/appveyor.yml

https://ci.appveyor.com/project/evanmiller/readstat/builds/21027690/job/d4og787nckjii33k/artifacts

However I don't know if that is deployed somewhere, maybe not.

Is that something that could help us? Like building ReadStat first with mingw, take the libs and then compile pyreadstat in MVSC linking against those libs?

**Comment by ofajardo** [comment]
Good news! There is already a ReadStat fork that supports MSVC. It was not merged back into ReadStat because ReadStat is built using autotools, which is not compatible with MSVC, so CI with appveyor fails. The discussion about what to do was stale for months, but it just happened last night that it was agreed that autotools could be replaced by CMake, and then ReadStat could be made compatible with MSVC. 

I guess this is going to take long time to be fully there, but once it is, for pyreadstat it will be super simple: following the current strategy of copying source files into the project, it should compile out of the box with MSVC and what you already have done in appveyor should start working right away.

Therefore, let's wait a bit for those changes and in the meantime I compile manually as previously discussed.

If you have any feedback from the mingw python guy, or any other comment, I would be interested on learning about it.

**Comment by ofajardo** [comment]
@matthew-brett as promised there is a new release now. Windows wheels are in the dist folder. 
Thanks a lot for taking care of this first build. I'll take care of the second one to get a grasp on the process, would be nice if I still can ask you questions if any. There is a lot of good improvements in ReadStat, so maybe there could be another release in late Jan or Feb. 
From there I can take care of this process regularly.

Regarding windows wheels, unfortunately pyreadstat does compile with python 2.7, but then test fails with an obscure error. Unfortunately I have limited time for this project, therefore I won't fix issues for python 2.7 (there is a chance that it is fixed once we are on MSVC, but it would be a side effect). I have written a disclaimer about that in the README.md. 
It works for me on python 2.7 in linux, I guess it will work on mac as well. In one hand I would drop wheels for python 2.7 on all systems to indicate it is not supported, in the other hand if they work I am OK with delivering them. Your opinion? I am fine with any of those.

Let me know how it goes.

**Comment by ofajardo** [comment]
_My own feeling is release the wheels unless they introduce a specific error (specific to the wheel). If they are going to get the same error compiling from source, why make them do that, to find out?_

fair point. I generated the wheel for python 2.7 and put it in the dist folder.

**Comment by matthew-brett** [comment]
I uploaded the wheels ...

**Comment by ofajardo** [comment]
Thanks! I put them in pypi tomorrow.

**Comment by matthew-brett** [comment]
Ah - sorry - I mean I uploaded the to PyPI - sorry if that's not what you wanted.

**Comment by matthew-brett** [comment]
No - you should be able to upload wheels for another platform without trouble.  Let me know?

**Comment by ofajardo** [comment]
Perfect!
For the next release: If in your repo I update the submodule pyreadstat, and then commit, will that be enough to trigger the built, meaning that let's say 20 min later the wheels will be in rackstack?
Or how is your recommended process?

**Comment by matthew-brett** [comment]
The process is in the README, somewhere, but the basic idea is:

* edit `.travis.yml` to set `BUILD_COMMIT`.
* commit, and push, to trigger the build.

There are also instructions there, for downloading from Rackspace, then uploading to PyPI, in a fairly automated way.

**Comment by ofajardo** [comment]
OK thanks, sounds pretty easy. I'll try next time.
